File: jugglerhelper/src/main/java/net/maxsmr/jugglerhelper/fragments/BaseJugglerFragment.java
package net.maxsmr.jugglerhelper.fragments;

import android.annotation.SuppressLint;
import android.app.Activity;
import android.content.Intent;
import android.graphics.drawable.Drawable;
import android.os.Build;
import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.support.annotation.CallSuper;
import android.support.annotation.ColorInt;
import android.support.annotation.LayoutRes;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentActivity;
import android.support.v4.content.ContextCompat;
import android.support.v4.util.Pair;
import android.view.KeyEvent;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MotionEvent;
import android.view.View;
import android.view.ViewGroup;
import android.view.Window;

import net.maxsmr.jugglerhelper.R;
import net.maxsmr.jugglerhelper.activities.BaseJugglerActivity;
import net.maxsmr.jugglerhelper.fragments.toolbar.BaseJugglerToolbarFragment;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.Collection;
import java.util.List;

import me.ilich.juggler.gui.JugglerActivity;
import me.ilich.juggler.gui.JugglerFragment;
import me.ilich.juggler.states.State;

public abstract class BaseJugglerFragment extends JugglerFragment {

    @NotNull
    protected final Handler mainHandler = new Handler(Looper.getMainLooper());

    @Nullable
    private Bundle savedInstanceState;

    private Menu menu;

    private boolean isCommitAllowed = false;

    public boolean isCommitAllowed() {
        return isAdded() && isCommitAllowed;
    }

    @Override
    protected BaseJugglerActivity getJugglerActivity() {
        return (BaseJugglerActivity) super.getJugglerActivity();
    }

    @Nullable
    public Fragment findChildFragmentById(int id) {
        Pair<Integer, Fragment> fragment = FragmentFinder.findFragmentById(getChildFragmentManager(), id);
        if (fragment != null) {
            return fragment.second;
        }
        return null;
    }

    @Nullable
    public Fragment findChildFragmentByTag(String tag) {
        Pair<Integer, Fragment> fragment = FragmentFinder.findFragmentByTag(getChildFragmentManager(), tag);
        if (fragment != null) {
            return fragment.second;
        }
        return null;
    }


    @Nullable
    public <F extends Fragment> F findChildFragmentByClass(Class<F> clazz) {
        Pair<Integer, F> fragment = FragmentFinder.findFragmentByClass(getChildFragmentManager(), clazz);
        if (fragment != null) {
            return fragment.second;
        }
        return null;
    }

    @Nullable
    public Fragment findRootFragmentById(int id) {
        return getJugglerActivity().findFragmentById(id);
    }

    @Nullable
    public Fragment findRootFragmentByTag(String tag) {
        return getJugglerActivity().findFragmentByTag(tag);
    }

    @Nullable
    public <F extends Fragment> F findRootFragmentByClass(Class<F> clazz) {
        return getJugglerActivity().findFragmentByClass(clazz);
    }

    @Nullable
    protected String getBaseFontAlias() {
        return null;
    }


    @Nullable
    public Bundle getSavedInstanceState() {
        return savedInstanceState;
    }

    @LayoutRes
    protected abstract int getContentLayoutId();

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        this.savedInstanceState = savedInstanceState;
        setHasOptionsMenu(false);
    }

    @Nullable
    @Override
    public final View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View rootView = null;
        int layoutId = getContentLayoutId();
        if (layoutId != 0) {
            rootView = inflater.inflate(layoutId, container, false);
        }
        if (rootView != null) {
            onBindViews(rootView);
        }
        return rootView;
    }

    protected abstract void onBindViews(@NotNull View rootView);

    protected Menu getMenu() {
        return menu;
    }

    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        this.menu = menu;
    }

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        isCommitAllowed = true;
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        setupWindow();
    }

    @Override
    public void onResume() {
        super.onResume();
        isCommitAllowed = true;
    }

    @Override
    public void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
        isCommitAllowed = false;
    }

    @CallSuper
    public void onTouchEvent(MotionEvent event) {
        if (isAdded()) {
            List<Fragment> childFragments = getChildFragmentManager().getFragments();
            for (Fragment f : childFragments) {
                if (f instanceof BaseJugglerFragment && f.isAdded()) {
                    ((BaseJugglerFragment) f).onTouchEvent(event);
                }
            }
        }
    }

    @CallSuper
    public void onKeyDown(int keyCode, KeyEvent e) {
        if (isAdded()) {
            List<Fragment> childFragments = getChildFragmentManager().getFragments();
            for (Fragment f : childFragments) {
                if (f instanceof BaseJugglerFragment && f.isAdded()) {
                    ((BaseJugglerFragment) f).onKeyDown(keyCode, e);
                }
            }
        }
    }

    @Override
    public boolean onBackPressed() {
        if (isAdded()) {
            List<Fragment> childFragments = getChildFragmentManager().getFragments();
            for (Fragment f : childFragments) {
                if (f instanceof BaseJugglerFragment && f.isAdded()) {
                    if (((BaseJugglerFragment) f).onBackPressed()) {
                        return true;
                    }
                }
            }
        }
        return false;
    }

    @CallSuper
    public void onStateActivated(JugglerActivity activity, State<?> state) {
        notifyStateActivated(activity, state, getChildFragmentManager().getFragments());
    }

    @Override
    @CallSuper
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        isCommitAllowed = true;
        notifyActivityResult(requestCode, resultCode, data, getChildFragmentManager().getFragments());
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NotNull String[] permissions, @NotNull int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        notifyRequestPermissionsResult(requestCode, permissions, grantResults, getChildFragmentManager().getFragments());
    }

    @SuppressWarnings("unchecked")
    @NotNull
    protected <T extends JugglerActivity> T getBaseActivity() {
        FragmentActivity activity = getActivity();

        if (activity == null) {
            throw new NullPointerException("not attached to activity");
        }
        if (!(activity instanceof JugglerActivity)) {
            throw new RuntimeException("activity isn't instance of a BaseActivity");
        }
        return (T) activity;
    }

    @Nullable
    protected Integer getScreenOrientation() {
        return null;
    }

    @Nullable
    @SuppressWarnings("ConstantConditions")
    protected Drawable getWindowBackground() {
        return null;
    }

    @Nullable
    @SuppressWarnings("ConstantConditions")
    @ColorInt
    protected Integer getStatusBarColor() {
        return ContextCompat.getColor(getContext(), R.color.colorStatusBar);
    }

    @Nullable
    @SuppressWarnings("ConstantConditions")
    @ColorInt
    protected Integer getNavigationBarColor() {
        return ContextCompat.getColor(getContext(), R.color.colorNavigationBar);
    }

    @SuppressLint("WrongConstant")
    protected void setupWindow() {
        final Activity activity = getActivity();
        if (activity != null) {
            final Integer orientation = getScreenOrientation();
            if (orientation != null) {
                activity.setRequestedOrientation(orientation);
            }
            final Window window = activity.getWindow();
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
                Integer statusBarColor = getStatusBarColor();
                if (statusBarColor != null) {
                    window.setStatusBarColor(statusBarColor);
                }
                Integer navigationBarColor = getNavigationBarColor();
                if (navigationBarColor != null) {
                    window.setNavigationBarColor(navigationBarColor);
                }
            }
            Drawable windowBackground = getWindowBackground();
            if (windowBackground != null) {
                window.setBackgroundDrawable(windowBackground);
            }
        }
    }

    protected void notifyStateActivated(JugglerActivity activity, State<?> state, @Nullable Collection<Fragment> fragments) {
        if (isAdded()) {
            if (fragments != null) {
                for (Fragment fragment : fragments) {
                    if (fragment != null && fragment.isAdded()) {
                        if (fragment instanceof BaseJugglerFragment) {
                            ((BaseJugglerFragment) fragment).onStateActivated(activity, state);
                        } else if (fragment instanceof BaseJugglerToolbarFragment) {
                            ((BaseJugglerToolbarFragment) fragment).onStateActivated(activity, state);
                        }
                    }
                }
            }
        }
    }

    protected void notifyActivityResult(int requestCode, int resultCode, Intent data, @Nullable Collection<Fragment> fragments) {
        if (isAdded()) {
            if (fragments != null) {
                for (Fragment f : fragments) {
                    if (f != null && f.isAdded())
                        f.onActivityResult(requestCode, resultCode, data);
                }
            }
        }
    }

    protected void notifyRequestPermissionsResult(int requestCode, @NotNull String[] permissions, @NotNull int[] grantResults, @Nullable Collection<Fragment> fragments) {
        if (isAdded()) {
            if (fragments != null) {
                for (Fragment f : fragments) {
                    if (f != null && f.isAdded()) {
                        f.onRequestPermissionsResult(requestCode, permissions, grantResults);
                    }
                }
            }
        }
    }
}

File: jugglerhelper/src/main/kotlin/net/maxsmr/jugglerhelper/fragments/BaseJugglerFragment.kt
package net.maxsmr.jugglerhelper.fragments

import android.annotation.SuppressLint
import android.content.Intent
import android.graphics.drawable.Drawable
import android.os.Build
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.view.*
import androidx.annotation.CallSuper
import androidx.annotation.ColorInt
import androidx.annotation.LayoutRes
import androidx.core.content.ContextCompat
import androidx.fragment.app.Fragment
import me.ilich.juggler.gui.JugglerActivity
import me.ilich.juggler.gui.JugglerFragment
import me.ilich.juggler.states.State
import net.maxsmr.jugglerhelper.R
import net.maxsmr.jugglerhelper.activities.BaseJugglerActivity
import net.maxsmr.jugglerhelper.fragments.toolbar.BaseJugglerToolbarFragment
import net.maxsmr.jugglerhelper.utils.FragmentSearchParams
import net.maxsmr.jugglerhelper.utils.findFragment

abstract class BaseJugglerFragment : JugglerFragment() {

    @get:LayoutRes
    protected abstract val layoutId: Int

    protected open val screenOrientation: Int? = null

    protected open val windowBackground: Drawable? = null

    protected open val statusBarColor: Int @ColorInt get() = ContextCompat.getColor(requireContext(), R.color.colorStatusBar)

    protected open val navigationBarColor: Int @ColorInt get()  = ContextCompat.getColor(requireContext(), R.color.colorNavigationBar)

    protected val mainHandler = Handler(Looper.getMainLooper())

    protected var savedInstanceState: Bundle? = null
        private set

    private var isCommitAllowed = false

    fun isCommitAllowed(): Boolean = isAdded && isCommitAllowed

    override fun getJugglerActivity(): BaseJugglerActivity =
            super.getJugglerActivity() as BaseJugglerActivity

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        this.savedInstanceState = savedInstanceState
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? =
            inflater.inflate(layoutId, container, false)

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        isCommitAllowed = true
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        setupWindow()
    }

    override fun onResume() {
        super.onResume()
        isCommitAllowed = true
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        isCommitAllowed = false
    }

    override fun onBackPressed(): Boolean {
        if (isAdded) {
            for (f in childFragmentManager.fragments) {
                if (f is BaseJugglerFragment && f.isAdded()) {
                    if (f.onBackPressed()) {
                        return true
                    }
                }
            }
        }
        return false
    }

    @CallSuper
    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        isCommitAllowed = true
        notifyActivityResult(requestCode, resultCode, data, childFragmentManager.fragments)
    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>, grantResults: IntArray) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        notifyRequestPermissionsResult(requestCode, permissions, grantResults, childFragmentManager.fragments)
    }

    @CallSuper
    open fun onStateActivated(activity: JugglerActivity?, state: State<*>?) {
        notifyStateActivated(activity, state, childFragmentManager.fragments)
    }

    @CallSuper
    fun onTouchEvent(event: MotionEvent?) {
        if (isAdded) {
            for (f in childFragmentManager.fragments) {
                if (f is BaseJugglerFragment && f.isAdded()) {
                    f.onTouchEvent(event)
                }
            }
        }
    }

    @CallSuper
    fun onKeyDown(keyCode: Int, e: KeyEvent?) {
        if (isAdded) {
            for (f in childFragmentManager.fragments) {
                if (f is BaseJugglerFragment && f.isAdded()) {
                    f.onKeyDown(keyCode, e)
                }
            }
        }
    }

    // TODO parentFragmentManager not found
//    fun <F : Fragment> findParentFragment(params: FragmentSearchParams<F>)= findFragment(parentFragmentManager, params)

    fun <F : Fragment> findChildFragment(params: FragmentSearchParams<F>)= findFragment(childFragmentManager, params)

    @Suppress("UNCHECKED_CAST")
    protected fun <P : State.Params> getParams(): P? {
        state?.let {
            return try {
                it.params as? P
            } catch (e: ClassCastException) {
                null
            }
        }
        return null
    }

    protected fun <P : State.Params> getParamsOrThrow(): P =
            getParams() ?: throw NullPointerException("Params not defined")

    @Suppress("UNCHECKED_CAST")
    protected fun <T : BaseJugglerActivity> getBaseActivity(): T =
            jugglerActivity as T

    @SuppressLint("WrongConstant")
    protected fun setupWindow() {
        with(requireActivity()) {
            val orientation = screenOrientation
            if (orientation != null) {
                requestedOrientation = orientation
            }
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
                statusBarColor.let {
                    if (it != 0) {
                        window.statusBarColor = it
                    }
                }
                navigationBarColor.let {
                    if (it != 0) {
                        window.navigationBarColor = it
                    }
                }
            }
            windowBackground?.let {
                window.setBackgroundDrawable(it)
            }
        }
    }

    protected fun notifyStateActivated(activity: JugglerActivity?, state: State<*>?, fragments: Collection<Fragment?>?) {
        if (isAdded) {
            if (fragments != null) {
                for (fragment in fragments) {
                    if (fragment != null && fragment.isAdded) {
                        if (fragment is BaseJugglerFragment) {
                            fragment.onStateActivated(activity, state)
                        } else if (fragment is BaseJugglerToolbarFragment) {
                            fragment.onStateActivated(activity, state)
                        }
                    }
                }
            }
        }
    }

    protected fun notifyActivityResult(requestCode: Int, resultCode: Int, data: Intent?, fragments: Collection<Fragment?>?) {
        if (isAdded) {
            if (fragments != null) {
                for (f in fragments) {
                    if (f != null && f.isAdded) {
                        f.onActivityResult(requestCode, resultCode, data)
                    }
                }
            }
        }
    }

    protected fun notifyRequestPermissionsResult(requestCode: Int, permissions: Array<String>, grantResults: IntArray, fragments: Collection<Fragment?>?) {
        if (isAdded) {
            if (fragments != null) {
                for (f in fragments) {
                    if (f != null && f.isAdded) {
                        f.onRequestPermissionsResult(requestCode, permissions, grantResults)
                    }
                }
            }
        }
    }
}

